### PR TITLE
[types/react-native] Bump to 0.71

### PR DESCRIPTION
**Pull request by lunaleaps**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/react-native/tree/0.71-stable/types
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

## Process
* Copied over type changes from 0.71-stable, including the folder structure
* Updated type libraries that relied on 0.70 types

### CODE REVIEW

**Comment by lunaleaps** [comment]
TODO Need to fix some naming discrepancies from running `npm run test-all`. I think we probably changed some names of types and maybe privatized some types that other libraries use.

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-native-vector-icons ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-native-vector-icons@latest/))
was missing the following properties:
1. createMultiStyleIconSet
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 92a5eff874b031588649ecbcc1c613a904944545
</p>

**Review by NickGerleman — APPROVED** [review]
Awesome work!

File: types/react-native-sensor-manager/index.d.ts
Comment (NickGerleman) [inline types/react-native-sensor-manager/index.d.ts:7]: I don't think the old version usage should be set here, since TS itself uses it for imports, and the local user won't have the same structure.

Comment (lunaleaps) [reply types/react-native-sensor-manager/index.d.ts:7]: I updated this due to this error: https://gist.github.com/lunaleaps/a184026d2c6cbeaa90f41f0c03f580f8
which doesn't seem to reference the `tsconfig.json` config for resolving `react-native` to the 0.70 version

Comment (lunaleaps) [reply types/react-native-sensor-manager/index.d.ts:7]: Seems like the issue is reported here: https://github.com/microsoft/TypeScript/issues/33206

Comment (lunaleaps) [reply types/react-native-sensor-manager/index.d.ts:7]: Removing it doesn't seem to impact the tests, `npm run react-native-sensor-manager` or `npm run test-all`, I think from the documentation triple-slash references usage should be dropped in favor of imports.. which the other DT packages seem to do

**Comment by lunaleaps** [comment]
I also realized I've accidentally commented out BatchedBridge, fixing

File: types/react-native-sensor-manager/index.d.ts
Comment (eps1lon) [reply types/react-native-sensor-manager/index.d.ts]: I think the `tsconfig.json` should resolve `react-native` to the 0.70 instead. That way the publisher also ensures `react-native-sensor-manager` declares a dependency on 0.70 types since it's clearly not compatible. 

Or fix the `react-native-sensor-manager` types to be compatible with latest.

**Review by eps1lon — CHANGES_REQUESTED** [review]
Just requesting changes so this isn't merged automatically considering 0.71 is not released yet. Just ping me once 0.71 is out and I'll merge ASAP.

File: types/react-native-sensor-manager/index.d.ts
Comment (lunaleaps) [reply types/react-native-sensor-manager/index.d.ts]: Yea, the `tsconfig.json` paths are ignored. So I'm removing the triple-slash reference

**Comment by kelset** [comment]
0.71 not out yet

**Comment by NickGerleman** [comment]
It seems like we’re going to potentially release 0.71 next week. I would normally be advocating for putting out the new typings *before* the release so that folks can update immediately. I still think it’s the best option, since folks can still edit the old types, and won’t automatically pull in 0.71 types. But since we’re already packaging the inbuilt types, DefinitelyTyped being out of date shouldn’t block people or force them to use incorrect types the same way it has before.

**Comment by eps1lon** [comment]
While this PR is released, new installs would be incorrect. And when RN release does get blocked, we have to revert and now have a bad release. Doing it after, like we do with every other package in the ecosystem, ensures all released types are sound. 

@lunaleaps PR needs rebase. If it's green I can merge immediately once RN 71 is out.

**Comment by NickGerleman** [comment]
> While this PR is released, new installs would be incorrect. And when RN release does get blocked, we have to revert and now have a bad release. Doing it after, like we do with every other package in the ecosystem, ensures all released types are sound.
> 
> @lunaleaps PR needs rebase. If it's green I can merge immediately once RN 71 is out.

Sure, waiting until the day of release seems like it would work (and I did hear of a delay). Though I don’t agree with several of the assertions here. 
1. Updating to “latest release + typings” without checking that major versions are available  is never a safe thing to do. Typings are often out of date compared to package release, and in this case, we could have the opposite. Though I could imagine people might still do this. 
2. Typings users targeting “^0.70” would continue getting 0.70 typings (semver considers the minor version as the major version when the major version is 0). 
3. We’ve had public RCs out for weeks which cannot use the correct copy of DefinitelyTyped typings because they haven’t been published. Though again, less of an issue now that we have the inbuilt ones, which we have been able to use during testing. 

I’ve been wondering though, if possible, if we want to go ahead and just mark the DefinitelyTyped package as deprecated when we push new typings. We don’t want folks to contribute to the typings here, since they are a mirror of the repo. And more specifically we want folks to get off of DefinitelyTyped before 0.72. Thoughts @lunaleaps @kelset ?

**Comment by kelset** [comment]
I don't have a strong opinion so I'll leave it to you and Luna to decide how you want to do about it; just a couple of things:

1. remember that DT has a specific approach for [deprecation/removal of types](https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package) that need to be followed
2. make sure that what you decide is correctly reflected in the blogpost about the TS types that will be published, currently it says a different thing from the one proposed

**Comment by eps1lon** [comment]
> Typings users targeting “^0.70” 

Remember that DT packages don't use SemVer ranges (but use `*`) and don't use peer dependencies. So any DT package importing from `react-native` will include the latest version of `@types/react-native` i.e. will break on new installs as I explained. 

I flagged this in https://github.com/microsoft/DefinitelyTyped-tools/issues/433 but it was rejected by the TypeScript team.

Then there's also packages wrongly using `>=` version ranges but then the responsibility is on them to ensure compatibility with all latest majors in perpetuity ;)

The correct solution would be to allow RCs of types but that was also ignored by TypeScript members (though no formal issue was opened).

**Comment by NickGerleman** [comment]
> I don't have a strong opinion so I'll leave it to you and Luna to decide how you want to do about it; just a couple of things:
> 
> 1. remember that DT has a specific approach for [deprecation/removal of types](https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package) that need to be followed
> 2. make sure that what you decide is correctly reflected in the blogpost about the TS types that will be published, currently it says a different thing from the one proposed

Oh interesting I hadn't known there was a built-in way to make DefinitelyTyped publish a stub redirecting to the package. It seems nice to, especially since 0.71 is just copying from the repo anyway, but I think this actually intersects with the DT packaging quirk @eps1lon mentioned:

> Remember that DT packages don't use SemVer ranges (but use *) and don't use peer dependencies. So any DT package importing from react-native will include the latest version of @types/react-native i.e. will break on new installs as I explained.

So a DefinitelyTyped library package depending on React Native will pull in and use the latest version of `@types/react-native` (I also verified this locally).

Even 2 months after 0.70 was released, 75% of weekly npm downloads went to older versions of react-native. So under the current scheme, it seems like these packages will almost always be using the incorrect version compared to the app. Regardless of when the @types/react-native updates would be released.

This isn't going to change the version of types used by application code which declares a dependency on a specific version, but means DT libraries would give incorrect types to the app (based on the types of a newer version of RN). Which seems to not blow up as much in practice now if it's almost aways slightly off. But moving those to stubs would have a side-effect that library typings would start pulling in a second whole copy of react-native 😕.

**Comment by lunaleaps** [comment]
I'm not sure I'm following entirely but have these thoughts/questions:

Re: deprecating DT
Won't we need DT types to live on for folks dependent on versions <0.71? We'll still need to release patch updates for those versions I'd imagine. Or maybe I'm misunderstanding what it means to "delete" types from DT

> But moving those to stubs would have a side-effect that library typings would start pulling in a second whole copy of react-native
Hmm okay so we're talking about libraries that depend on react-native and have DT types themselves. I'm wondering how common that is and could also publish types as a separate package from `react-native`?

**Review by eps1lon — APPROVED** [review]

**Review by kelset — APPROVED** [review]

**Comment by lunaleaps** [comment]
Ready to merge

**Pull request merged**
